imports/client/hooks/useCallState.ts

```diff
@@ -0,0 +1,789 @@
+      const method = direction === 'send' ? 'createSendTransport' : 'createRecvTransport';
+      const newTransport = device[method]({
+        id: transportId,
+        iceParameters: JSON.parse(iceParameters),
+        iceCandidates: JSON.parse(iceCandidates),
+        dtlsParameters: JSON.parse(serverDtlsParameters),
+        appData: {
+          _id,
+        },
+      });
+      newTransport.on('connect', ({ dtlsParameters: clientDtlsParameters }, callback) => {
+        connectRef.current = callback;
+        // No need to set a callback here, since the ConnectAck record acts as a
+        // callback
+        mediasoupConnectTransport.call({
+          transportId: _id,
+          dtlsParameters: JSON.stringify(clientDtlsParameters),
+        }, (err) => {
+          if (err) {
+            console.error(`Failed to connect transport ${direction}`, err);
+          }
+        });
+      });
+      log(`setting ${direction} transport`, newTransport);
+      dispatch({
+        type: 'set-transport',
+        direction,
+        transport: newTransport,
+      });
+      return () => {
+        if (!newTransport.closed) newTransport.close();
+      };
+    } else {
+      log(`clearing ${direction} transport`);
+      dispatch({
+        type: 'set-transport',
+        direction,
+        transport: undefined,
+      });
+      return undefined;
+    }
+  }, [
+    device, hasParams, transportParams?._id, direction,
+    transportParams?.transportId, transportParams?.iceParameters, transportParams?.iceCandidates,
+    transportParams?.dtlsParameters, dispatch,
+  ]);
+
+  return connectRef;
+};
+
+type ProducerCallback = ({ id }: { id: string }) => void;
+type ProducerState = {
+  producer: types.Producer | undefined;
+  subHandle: Meteor.SubscriptionHandle | undefined;
+  producerServerCallback: ProducerCallback | undefined;
+  kind: string | undefined;
+  rtpParameters: string | undefined;
+}
+
+type ConsumerState = {
+  consumer: types.Consumer | undefined;
+  peerId: string;
+}
+
+const useCallState = ({ huntId, puzzleId, tabId }: {
+  huntId: string,
+  puzzleId: string,
+  tabId: string,
+}): [CallState, React.Dispatch<Action>] => {
+  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
+
+  useEffect(() => {
+    // If huntId, puzzleId, or tabId change (but mostly puzzleId), reset
+    // call state.
+    return () => {
+      log('huntId/puzzleId/tabId changed, resetting call state');
+      dispatch({ type: 'reset' });
+    };
+  }, [huntId, puzzleId, tabId]);
+
+  useEffect(() => {
+    // When mediaSource (the mic capture stream) changes, stop all the tracks.
+    return () => {
+      if (state.audioState?.mediaSource) {
+        state.audioState?.mediaSource.getTracks().forEach((track) => {
+          track.stop();
+        });
+      }
+    };
+  }, [state.audioState?.mediaSource]);
+
+  const joinSubRef = useRef<Meteor.SubscriptionHandle | undefined>(undefined);
```

Comment:
The reason we handle this subscription (and the `mediasoup:transports` sub below) manually instead of using `useTracker` or `useSubscription` is because of an unfortunate interaction between:

* us using these publishes to cause server-side state changes
* us expecting `[huntId, puzzleId, tabId]` (or below, `peer._id`) to generally be unique per client, with last-writer-wins semantics on conflict
* React reserving the right to rerun `render()` multiple times, as exercised by `StrictMode`
* `useTracker` (and as a result `useSubscription`) running the inner function body during both the render phase and the effect phase, resulting in multiple subscriptions

The solution is to move our use of state-changing subscriptions to run exclusively in the effect phase, which is reasonably straightforward.

In hindsight, this probably deserves a comment.